### ISSUE TITLE
[Rock-ons] Fix unused container_links object in JSON definition file

### ISSUE THREAD

**Issue by FroggyFlox** [rockstor/rockstor-core#2003]
The current `container_links` object in a Rock-on definition file is currently used only by an older (and seemingly unused?) custom function for owncloud install. Furthermore, as brought up by @Luke-Nukem in a separate issue (#1982), container links are being deprecated by docker in favor of use networks.

Following discussion in this other issue (#1982), we can implement the following using the current syntax of `container_links` and docker networks, as suggested and detailed by @Luke-Nukem:
https://github.com/rockstor/rockstor-core/issues/1982#issue-374932916

The suggestion is repasted below for reference:

The best thing to do would be to retain the existing container linking json, and use that to determine how to network:
```
    "container_links": {
      "container-A": [
        {
          "name": "link-B",
          "source_container": "container-B"
        },
        {
          "name": "link-C",
          "source_container": "container-C"
        }
      ]
    }
,
```
Using the above A needs to see B and C, but B and C don't need to see each other. The docker commands would simply be:
```
docker network create link-B
docker network connect container-A link-B
docker network connect container-B link-B
// create network if not exists
docker network create link-C
docker network connect container-A link-C
docker network connect container-C link-C
```

I'll start working on this issue.

**Comment by Luke-Nukem** [comment]
Brilliant! I look forward to seeing how it pans out.

**Comment by FroggyFlox** [comment]
Just a quick update, @Luke-Nukem , I now have a working implementation of your solution based on the existing `container_links` JSON object.
I tested it with the same conditions as you described so far (3 containers connected separately or together) and is behaving as expected. I still need to test it with more complex situations, so feel free to describe a complex scenario if you have one in mind.
Once I've done proper testing, I'll proceed further with the PR submission.

Cheers,

**Comment by FroggyFlox** [comment]
The current logic seems to be able to deal with the scenarios I've tested so far so I'll detail this here.

### Current shortcoming
As we're keeping the same `container_links` object in the JSON file, I used its currently established parsing to the `DContainerLink` storageadmin table, hoping we wouldn't have to modify it in any way. Even though this works perfectly for separate networks within a Rock-on, this unfortunately cannot support networks sharing a destination container and a unique `destination container + network name` is currently required. Expanding this requirement to `source container + destination container + network name` resolves the issue and allows for one container to connect to multiple containers using the same network.

### Logic and implementation
1. Start dealing with docker networks after creation and start of docker containers, as connecting a container to a network requires the container to exist.
2. For each container in the Rock-on, check for an existing link definition and proceed to link creation and container connection for each link.
3. Check if link already exist and create if needed.
4. Check if destination container is running (or at least not absent) and connect it to the given network if not already in it.
5. Repeat step4 for the source container
6. Upon Rock-on uninstall loop trough all links associated to the container(s) in the Rock-on and remove the corresponding network(s).

In addition to having been tested with the links described in the original post, this logic has been verified to work with a test Rock-on (json file [here](https://gist.github.com/FroggyFlox/a0ac18c82cb1e8926fe3c4e8825e034d)). This Rock-on runs 4 copies of the alpine image, with the following links, repasted below:
- A sees B
- C sees D
- A sees C
```
	"Alpine_link_test_ABCD_2nets": {
		"container_links": {
            "alpineA": [
                {
                    "name": "alpineA-B",
                    "source_container": "alpineB"
                },
				{
                    "name": "alpineA-C",
                    "source_container": "alpineC"
                }
            ],
			"alpineC": [
				{
                    "name": "alpineC-D",
                    "source_container": "alpineD"
                }
			]
        },
```


As expected, this Rock-on installs successfully and creates three new networks:
```
[root@rockdev ~]# docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
4d5427ff0554        alpineA-B           bridge              local
802438d35617        alpineA-C           bridge              local
070688a3b98a        alpineC-D           bridge              local
dd12f59401b6        bridge              bridge              local
1aa5ee642881        host                host                local
eeb108871876        none                null                local
```

Each network connects the correct containers:
```
[root@rockdev ~]# docker network inspect -f '{{range .Containers}}{{.Name}}{{end}}' alpineA-C
alpineCalpineA
[root@rockdev ~]# docker network inspect -f '{{range .Containers}}{{.Name}}{{end}}' alpineA-B
alpineAalpineB
[root@rockdev ~]# docker network inspect -f '{{range .Containers}}{{.Name}}{{end}}' alpineA-C
alpineCalpineA
[root@rockdev ~]# docker network inspect -f '{{range .Containers}}{{.Name}}{{end}}' alpineC-D
alpineCalpineD
```

And each container can ping the correct ones:
```
[root@rockdev ~]# docker attach alpineA
/ # for C in 'alpineA' 'alpineB' 'alpineC' 'alpineD'; do ping -c 2 $C ; done
PING alpineA (172.18.0.2): 56 data bytes
64 bytes from 172.18.0.2: seq=0 ttl=64 time=0.029 ms
64 bytes from 172.18.0.2: seq=1 ttl=64 time=0.057 ms

--- alpineA ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.029/0.043/0.057 ms
PING alpineB (172.18.0.3): 56 data bytes
64 bytes from 172.18.0.3: seq=0 ttl=64 time=0.063 ms
64 bytes from 172.18.0.3: seq=1 ttl=64 time=0.150 ms

--- alpineB ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.063/0.106/0.150 ms
PING alpineC (172.19.0.3): 56 data bytes
64 bytes from 172.19.0.3: seq=0 ttl=64 time=0.061 ms
64 bytes from 172.19.0.3: seq=1 ttl=64 time=0.148 ms

--- alpineC ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.061/0.104/0.148 ms
ping: bad address 'alpineD'
/ # read escape sequence
[root@rockdev ~]# 
[root@rockdev ~]# 
[root@rockdev ~]# docker attach alpineB
/ # 
/ # 
/ #  for C in 'alpineA' 'alpineB' 'alpineC' 'alpineD'; do ping -c 2 $C ; done
PING alpineA (172.18.0.2): 56 data bytes
64 bytes from 172.18.0.2: seq=0 ttl=64 time=0.173 ms
64 bytes from 172.18.0.2: seq=1 ttl=64 time=0.232 ms

--- alpineA ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.173/0.202/0.232 ms
PING alpineB (172.18.0.3): 56 data bytes
64 bytes from 172.18.0.3: seq=0 ttl=64 time=0.225 ms
64 bytes from 172.18.0.3: seq=1 ttl=64 time=0.127 ms

--- alpineB ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.127/0.176/0.225 ms
ping: bad address 'alpineC'
ping: bad address 'alpineD'
/ # read escape sequence
[root@rockdev ~]# 
[root@rockdev ~]# 
[root@rockdev ~]# docker attach alpineC
/ #  for C in 'alpineA' 'alpineB' 'alpineC' 'alpineD'; do ping -c 2 $C ; done
PING alpineA (172.19.0.2): 56 data bytes
64 bytes from 172.19.0.2: seq=0 ttl=64 time=2.033 ms
64 bytes from 172.19.0.2: seq=1 ttl=64 time=0.150 ms

--- alpineA ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.150/1.091/2.033 ms
ping: bad address 'alpineB'
PING alpineC (172.19.0.3): 56 data bytes
64 bytes from 172.19.0.3: seq=0 ttl=64 time=0.052 ms
64 bytes from 172.19.0.3: seq=1 ttl=64 time=0.218 ms

--- alpineC ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.052/0.135/0.218 ms
PING alpineD (172.20.0.3): 56 data bytes
64 bytes from 172.20.0.3: seq=0 ttl=64 time=0.294 ms
64 bytes from 172.20.0.3: seq=1 ttl=64 time=0.065 ms

--- alpineD ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.065/0.179/0.294 ms
/ # read escape sequence
[root@rockdev ~]# 
[root@rockdev ~]# 
[root@rockdev ~]# docker attach alpineD
/ #  for C in 'alpineA' 'alpineB' 'alpineC' 'alpineD'; do ping -c 2 $C ; done
ping: bad address 'alpineA'
ping: bad address 'alpineB'
PING alpineC (172.20.0.2): 56 data bytes
64 bytes from 172.20.0.2: seq=0 ttl=64 time=0.063 ms
64 bytes from 172.20.0.2: seq=1 ttl=64 time=0.097 ms

--- alpineC ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.063/0.080/0.097 ms
PING alpineD (172.20.0.3): 56 data bytes
64 bytes from 172.20.0.3: seq=0 ttl=64 time=0.149 ms
64 bytes from 172.20.0.3: seq=1 ttl=64 time=0.102 ms

--- alpineD ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.102/0.125/0.149 ms
/ # read escape sequence
```

After Rock-on uninstall, all networks are correctly deleted:
```
[root@rockdev ~]# docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
dd12f59401b6        bridge              bridge              local
1aa5ee642881        host                host                local
eeb108871876        none                null                local
```

As mentioned above, this solution seems to work but as it requires an update to one of Rockstor's database definitions, I'll wait until the current PR are merged before submitting a corresponding PR (as this would require a rebase anyway). This does mean we can further test and improve the current system until then.

In advance, thanks for any feedback on this.

**Comment by phillxnet** [comment]
@FroggyFlox This is an impressive development and very well reported. Definitely raises the Rock-on capability and I like how you have re-purposed the existing mechanism. To your question:

"... do you think I should try to have the docker network name be displayed instead?"

Yes I do think it would make sense as then we feed back to the user what they named the networks in the first place. Otherwise it's prone to confusion, ie "what are all these networks and where did they come from." You could have a method within the network code that simply builds a dictionary from parsing the output of 'docker network ls' and builds this dict which in turn allows for populating that table with the 'friendly' names. It may even be worth adding another column for this name, that way you don't have to mess with what is obviously working as is.

Your call of course, and this could always be added at a later date via a separate issue / pr, but I'd vote for having the friendly name displayed if it was doable as then it nicely rounds out from a UI perspective. But then it's not me doing it.

As per additional db changes, I say go for it. If we need them then so be it. I have 3 more to add myself once your current pending pr is merged so we might as well continue our run of db changes and aim for a calmer period there after. I'll try and pop my pr in once your last one is done so we will just have to watch each others submissions as and when. I have a little writing up to do first so you may end up beating me to it. We can always pm to try and co-ordinate anyway.

Great work by the way, this has all come along faster than I imagined it would so well done. And thanks for your efforts in it's explanation and function testing, I'm afraid it's another weak area for me so this helps. And feel free to ignore my glib suggestion if it really isn't as simple as just 'building a dictionary and adding a column' :).

**Comment by FroggyFlox** [comment]
Thanks @phillxnet for the feedback, but all credit goes to @Luke-Nukem as I simply implemented his idea of using the existing `container_links` and as all the JSON parsing to database was already in place, I only had to write down the docker calls.

I'll work on displaying the actual docker network names in addition to the ids as it currently is the case. Even though the user doesn't choose these names so far (they're defined in the Rock-on JSON), it's a lot more clear. Plus, it will become necessary once we implement the remaining options that will allow the user to create and join their docker user network of choice. I will follow your suggestion on building a dictionary from `docker network ls`. There's already a little bit of that in the current code, but that will definitively be a very useful method in the implementation of the remaining docker networks options. Thanks!

I think I prefer to implement this before submitting a PR, actually, given the improvement it represents and the current dependency on a pending PR. Given this one will involve work on the UI side of Rockstor, it'll take me more time to finish anyway so the current pending PRs and your coming one(s) will most likely be there before that.

**Comment by Luke-Nukem** [comment]
> _Should these docker-related networks created from Rock-ons definition files or Docker defaults be allowed to be edited / deleted by the user from the web-UI?_

I don't see any value to this. The point of a Docker compose or Rock-on is to be a self contained setup that requires no user intervention. You run the compose, and everything is magicked in to existence.

Having said that, when it comes to something like an Nginx container for routing/ssh, then that needs to be added to the networks the user wishes to be able to route ports/domains/etc on.

For example, my NAS currently runs an Owncloud compose, plus a Gitea compose, and I use Nginx as an SSL proxy to these which are on two separate networks.

There is perhaps value in:
1. having the ability for users to add one container to another network,
2. create new networks and add containers to them,

Though anybody with the requirement to muck about with docker networking would probably be able to do it through SSH also. The Nginx example above requires a bit of config tweaking, so unless this was automated then network adjustments aren't really something a user would be doing. Having said that, a more powerful addition to Rockstor later may be that ability - use of Nginx to put non-HTTPS containers behind SSL.

Example nginx config:
```
user  nginx;
worker_processes  2;

error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;

events {
  worker_connections   512;
  # use [ kqueue | rtsig | epoll | /dev/poll | select | poll ] ;
  use epoll;
}

http {
    include         /etc/nginx/mime.types;
    default_type    application/octet-stream;

    client_header_timeout   3m;
    client_body_timeout     3m;
    send_timeout            3m;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log      /var/log/access.log  main;

    gzip on;
    gzip_min_length 1100;
    gzip_buffers    4 8k;
    gzip_types      text/plain;

    sendfile        on;
    tcp_nopush      on;

    tcp_nodelay     on;
    
    keepalive_timeout  65;

    proxy_redirect   off;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host $server_name;

    include     /etc/nginx/ssl.conf;

    # Base server
    server {
        listen      443 ssl;
        listen      80 default_server;
        listen      [::]:80 default_server;  
        server_name home.ljcode.org www.home.ljcode.org blog.ljcode.org;
        access_log  /var/log/access.home.ljcode.org.log  main;
        location / {
            root   /usr/share/nginx/html;
            index  index.html index.htm;
        }

        location /owncloud {
            rewrite ^/owncloud(.*) /$1 break;
            proxy_pass http://owncloud_server_1:8080;
        }
    }

    # Proxy to owncloud_server_1 docker for SSL
    server {
        listen      443 ssl http2;
        listen      [::]:443 ssl http2;
        listen      80;
        listen      8080;
        listen      [::]:80;
        listen      [::]:8080;
        server_name cloud.ljcode.org www.cloud.ljcode.org 192.168.20.2;
        access_log  /var/log/access.owncloud_server_1.log  main;
        client_max_body_size 1000M; # required for file sizes
        location / {
            proxy_pass http://owncloud_server_1:8080;
        }
    }

#    # Proxy to gitea docker for SSL
    server {
        listen      443 ssl http2;
        listen      [::]:443 ssl http2;
        listen      80;
        listen      [::]:80;
        listen      3000;
        listen      [::]:3000;
        server_name git.ljcode.org www.git.ljcode.org;
        access_log  /var/log/access.gitea.log  main;
        location / {
            proxy_pass http://gitea_server_1:3000;
        }
    }
}
stream {
    proxy_connect_timeout 10s;
    proxy_timeout 1m;
#    # HTTPS pass-through for gitea
#    # Doesn't work well for cloning via https
#    server {
#        listen 6661;
#        proxy_pass gitea:3000;
#    }
    # SSH pass-through for gitea
    server {
        listen 6662;
        proxy_pass gitea_server_1:22;
    }
}
```

**Comment by FroggyFlox** [comment]
Hi @Luke-Nukem and thanks again for your feedback.

> I don't see any value to this.

I agree with you and was leaning towards removing the option for deactivating. editing, or deleting the connection for these docker networks (those defined in a Rock-on JSON file). They are necessary for the Rock-on proper function and do not need to be altered by the user. 

This would be in contrast to docker networks created by the user, however. Indeed, as discussed in #1982, giving the user a simple interface to create a docker network and attach any running container(s) of interest (addressing the two points you listed). As these networks can be customized through `docker network` tools (subnet, IP, gateways, etc...), I will work on offering these "advanced" options (offered only to those who seek them) for them. Even though these can't be modified on the fly through docker tools, I will see if I can find a way to do so through other ways.

> Having said that, a more powerful addition to Rockstor later may be that ability - use of Nginx to put non-HTTPS containers behind SSL.

I agree with that too. It's actually something I had in mind for a little while now and one of the reasons why I wanted to add support for container labels (PR #1999) in the prospect of using Traefik for this exact purpose. As placing a container behind it is mostly based on using labels, I wanted to design a system to simplify exactly that.

I'm still quite a way from that, however, and need to finish the steps detailed in #1982 first. I'm going to start working on a way to "un-export" ports and then creating custom networks.

**Comment by FroggyFlox** [comment]
_Quick update:_
Following the discussion above, all edits and deletion of docker netowrks _created from Rock-on JSON in_ `container_links` _objects_ are deactivated.
To do so, I implemented a way to detect whether a docker network detected by `nmcli` corresponds to a network created from a `container_links` object, or to a docker network created by the user. This thus allows to offer the user with an GUI to edit or delete properly docker networks that are user-defined, and will be helpful for the remaining steps of #1982. 

In the illustration below, _alpineX-X_ connections are docker networks created from `container_links`, whereas _testnet01_ corresponds to a docker network created by the user (from the cli so far).

![image](https://user-images.githubusercontent.com/30297881/50846993-73033180-133e-11e9-918b-eb593891cec3.png)
